File: Functions/plotNyquistDiagram.py
"""
plotNyquistDiagram - Plot Nyquist diagram from impedance data

This function plots a Nyquist diagram from the provided data DataFrame.
It automatically detects different SoC tests based on column naming
convention (_SoC##) and creates separate legend entries.

Author: Róbinson Medina, Feye Hoekstra
Updated: February 12, 2026
"""

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import re


def plot_nyquist_diagram(data, plot_title='Nyquist Diagram', save_path='', max_freq=12000):
    """
    Plot Nyquist diagram from impedance data.
    
    Parameters
    ----------
    data : pandas.DataFrame
        DataFrame containing R_real_ohm_SoC##, R_img_ohm_SoC##, and 
        Freq_Hz_SoC## columns for different SoC values
    plot_title : str, optional
        Custom title for the plot (default: 'Nyquist Diagram')
    save_path : str, optional
        Path to save the figure (empty string to skip saving)
    max_freq : float, optional
        Maximum frequency for data inclusion in Hz (default: 12000)
    
    Returns
    -------
    fig : matplotlib.figure.Figure
        Figure handle
    """
    
    # Get column names
    col_names = data.columns.tolist()
    
    # Find unique SoC values by parsing column names
    soc_values = []
    for col in col_names:
        if 'R_real_ohm_SoC' in col:
            # Extract SoC value from column name
            match = re.search(r'SoC(\d+)', col)
            if match:
                soc_str = match.group(1)
                if soc_str not in soc_values:
                    soc_values.append(soc_str)
    
    if not soc_values:
        raise ValueError('No SoC data found. Expected columns with format R_real_ohm_SoC##')
    
    print(f'Found {len(soc_values)} different SoC test(s): {", ".join(soc_values)}')
    
    # Create figure
    fig, ax = plt.subplots(figsize=(10, 8))
    
    # Define colors for different SoC values
    colors = plt.cm.tab10(np.linspace(0, 1, len(soc_values)))
    
    # Plot data for each SoC
    legend_entries = []
    for i, soc_val in enumerate(soc_values):
        # Column names for this SoC
        r_real_col = f'R_real_ohm_SoC{soc_val}'
        r_img_col = f'R_img_ohm_SoC{soc_val}'
        freq_col = f'Freq_Hz_SoC{soc_val}'
        
        # Check if columns exist
        if r_real_col not in col_names or r_img_col not in col_names:
            print(f'Warning: Missing columns for SoC{soc_val}, skipping...')
            continue
        
        # Extract data
        r_real = data[r_real_col].values
        r_img = data[r_img_col].values
        
        # Get frequency data for filtering
        freq = None
        if freq_col in col_names:
            freq = data[freq_col].values
        
        # Remove NaN values and apply frequency filter
        if freq is not None:
            valid_idx = ~np.isnan(r_real) & ~np.isnan(r_img) & ~np.isnan(freq) & (freq <= max_freq)
        else:
            valid_idx = ~np.isnan(r_real) & ~np.isnan(r_img)
        
        r_real = r_real[valid_idx]
        r_img = r_img[valid_idx]
        if freq is not None:
            freq = freq[valid_idx]
        
        if len(r_real) == 0:
            print(f'Warning: No valid data for SoC{soc_val}, skipping...')
            continue
        
        # Plot Nyquist curve (-r_img on y-axis for conventional Nyquist representation)
        ax.plot(r_real, -r_img, 'o-', color=colors[i], linewidth=1.5,
                markersize=4, markerfacecolor=colors[i], label=f'SoC {soc_val}%')
        
        legend_entries.append(f'SoC {soc_val}%')
    
    # Format plot
    ax.set_xlabel(r'$R_{real}$ ($\Omega$)', fontsize=12)
    ax.set_ylabel(r'$-R_{img}$ ($\Omega$)', fontsize=12)
    ax.set_title(plot_title, fontsize=14, fontweight='bold')
    
    # Add legend
    if legend_entries:
        ax.legend(loc='best', fontsize=10)
    
    # Make axes equal for proper Nyquist representation
    ax.set_aspect('equal', adjustable='box')
    
    # Add grid
    ax.grid(True, which='both', linestyle='-', alpha=0.7)
    ax.minorticks_on()
    ax.grid(True, which='minor', linestyle=':', alpha=0.4)
    
    # Improve appearance
    ax.tick_params(labelsize=10)
    
    plt.tight_layout()
    
    # Save figure if requested
    if save_path:
        fig.savefig(save_path, dpi=150, bbox_inches='tight')
        print(f'Nyquist diagram saved as: {save_path}')
    
    print(f'Nyquist diagram created successfully with {len(legend_entries)} SoC test(s)')
    
    return fig

File: Functions/plotNyquistDiagram.m
function fig = plotNyquistDiagram(data, plotTitle, savePath, maxFreq)
%PLOTNYQUISTDIAGRAM Plot Nyquist diagram from impedance data
%   fig = plotNyquistDiagram(data, plotTitle, savePath, maxFreq) plots a 
%   Nyquist diagram from the provided data table. The function automatically 
%   detects different SoC tests based on column naming convention (_SoC##) 
%   and creates separate legend entries.
%
%   Input:
%       data - Table containing R_real_ohm_SoC##, R_img_ohm_SoC##, and 
%              Freq_Hz_SoC## columns for different SoC values
%       plotTitle - Custom title for the plot
%       savePath - Path to save the figure (empty string to skip saving)
%       maxFreq - Maximum frequency for data inclusion (Hz)
%
%   Output:
%       fig - Figure handle
%
%   Example:
%       fig = plotNyquistDiagram(myData, 'My Nyquist Plot', '', 12000);

    % Set default values if not provided
    if nargin < 2 || isempty(plotTitle)
        plotTitle = 'Nyquist Diagram';
    end
    if nargin < 3
        savePath = '';
    end
    if nargin < 4 || isempty(maxFreq)
        maxFreq = 12000;
    end
    
    % Get column names
    colNames = data.Properties.VariableNames;
    
    % Find unique SoC values by parsing column names
    socValues = {};
    for i = 1:length(colNames)
        if contains(colNames{i}, 'R_real_ohm_SoC')
            % Extract SoC value from column name
            socStr = extractAfter(colNames{i}, 'SoC');
            if ~ismember(socStr, socValues)
                socValues{end+1} = socStr;
            end
        end
    end
    
    if isempty(socValues)
        error('No SoC data found. Expected columns with format R_real_ohm_SoC##');
    end
    
    fprintf('Found %d different SoC test(s): %s\n', length(socValues), strjoin(socValues, ', '));
    
    % Create figure
    fig = figure('Position', [100, 100, 800, 600]);
    hold on;
    grid on;
    
    % Define colors for different SoC values
    colors = lines(length(socValues));
    
    % Plot data for each SoC
    legendEntries = {};
    for i = 1:length(socValues)
        socVal = socValues{i};
        
        % Column names for this SoC
        rRealCol = ['R_real_ohm_SoC' socVal];
        rImgCol = ['R_img_ohm_SoC' socVal];
        freqCol = ['Freq_Hz_SoC' socVal];
        
        % Check if columns exist
        if ~ismember(rRealCol, colNames) || ~ismember(rImgCol, colNames)
            fprintf('Warning: Missing columns for SoC%s, skipping...\n', socVal);
            continue;
        end
        
        % Extract data
        rReal = data.(rRealCol);
        rImg = data.(rImgCol);
        
        % Get frequency data for filtering
        freq = [];
        if ismember(freqCol, colNames)
            freq = data.(freqCol);
        end
        
        % Remove NaN values and apply frequency filter
        if ~isempty(freq)
            validIdx = ~isnan(rReal) & ~isnan(rImg) & ~isnan(freq) & freq <= maxFreq;
        else
            validIdx = ~isnan(rReal) & ~isnan(rImg);
        end
        
        rReal = rReal(validIdx);
        rImg = rImg(validIdx);
        if ~isempty(freq)
            freq = freq(validIdx);
        end
        
        if isempty(rReal)
            fprintf('Warning: No valid data for SoC%s, skipping...\n', socVal);
            continue;
        end
        
        % Plot Nyquist curve
        h = plot(rReal, -rImg, 'o-', 'Color', colors(i,:), 'LineWidth', 1.5, ...
             'MarkerSize', 4, 'MarkerFaceColor', colors(i,:));
        
        % Add custom datatip with impedance labels
        h.DataTipTemplate.DataTipRows(1).Label = 'Real impedance';
        h.DataTipTemplate.DataTipRows(1).Value = rReal;
        h.DataTipTemplate.DataTipRows(2).Label = 'Imaginary impedance';
        h.DataTipTemplate.DataTipRows(2).Value = -rImg;
        % xlim([-0 4.5e-3]);
        % ylim([-4.5e-3 1.5e-3]);
        
        % Add frequency information if available
        if ismember(freqCol, colNames) && ~isempty(freq)
            row = dataTipTextRow('Frequency (Hz)', freq);
            h.DataTipTemplate.DataTipRows(end+1) = row;
        end
        
        % Add to legend
        legendEntries{end+1} = ['SoC ' socVal '%'];
    end
    
    % Format plot
    xlabel('R_{real} (\Omega)', 'FontSize', 12);
    ylabel('-R_{img} (\Omega)', 'FontSize', 12);
    title(plotTitle, 'FontSize', 14, 'FontWeight', 'bold', 'Interpreter', 'none');
    
    % Add legend
    if ~isempty(legendEntries)
        legend(legendEntries, 'Location', 'best', 'FontSize', 10);
    end
    
    % Make axes equal for proper Nyquist representation
    axis equal;
    
    % Add grid
    grid on;
    grid minor;
    
    % Improve appearance
    set(gca, 'FontSize', 10);
    box on;
    
    hold off;
    
    % Save figure if requested
    if ~isempty(savePath)
        saveas(fig, savePath, 'png');
        fprintf('Nyquist diagram saved as: %s\n', savePath);
    end
    
    fprintf('Nyquist diagram created successfully with %d SoC test(s)\n', length(legendEntries));
end
